base/file.jl

```diff
@@ -356,13 +339,33 @@ is an open file object for this path.
 mktemp(parent)
 
 """
-    mktempdir(parent=tempdir())
+    mktempdir(parent=tempdir(); prefix="$temp_prefix")
 
 Create a temporary directory in the `parent` directory and return its path.
-If `parent` does not exist, throw an error.
+If `parent` does not exist, throw an error. An optional `prefix` to the directory name can
+be provided.
 """
-mktempdir(parent)
+function mktempdir(parent=tempdir(); prefix="$temp_prefix")
+    template = prefix*"XXXXXX"
+    tpath = joinpath(parent, template)
```

Comment:
I think the intention of this API was to compute simply:
`tpath = "$(parent)$(path_separator)$(template)XXXXXX"`

Reply:
You mean that it's simpler/more efficient to allocate the string in a single pass?

Reply:
That doesn't matter here. However, the goal of the API appears to be to compute the simpler operation, not joinpath.

Reply:
`joinpath` has 2 functions:

1. Use the relevant $(path_separator)
2. Provide a sane path relevant to the OS by stripping end separator. 
```
     ismatch(path_separator_re, a[end:end]) ? string(C,a,b) :
                                             string(C,a,pathsep(a,b),b)
```

If the input to `parent` variable ends with a `$(path_separator)` then the path created will have 2 consecutive `$path_separator` values. It does not fail but is not the correct return value. 

Example:
```
julia> mktempdir("/tmp/"; prefix="abc")
"/tmp//abclEMi2T"

julia> mktempdir("/"; prefix="tmp/")
"//tmp/rXRggu"
```

Reply:
@vtjnash Can you review and suggest if `joinpath` should be removed from the code.

Reply:
Yes, it should. But to your point, yes we can add some logic here to avoid duplicating the path separator if already present at the end of parent:
```
if ismatch(path_separator_re, parent[end:end])
    tpath = "$(parent)$(template)XXXXXX"
else
    tpath = "$(parent)$(path_separator)$(template)XXXXXX"
end
```